feat(examples): add {p,r}/agherasie/forms

agherasie opened the pull request:
Sorry for the long hiatus which caused #2604 to close, I moved to South Korea this year for a university exchange program and let this PR collect dust for a while.
I've addressed https://github.com/gnolang/gno/pull/2604#discussion_r1787776352 in https://github.com/gnolang/gno/commit/7a6a03235e1a615d2fa942294ac625a170de6ba9, hoping that the PR might be ready for merge now !
If not, please let me know if any further updates should be made to the code !

---

As part of the student contributor program, I attempted to create a new example realm that allows the creation and submission of forms on gno !

## Features
- **Form Creation**: Create new forms with specified titles, descriptions, and fields.
`CreateForm(...)`
- **Form Submission**: Submit answers to forms.
`SubmitForm(...)`
- **Form Retrieval**: Retrieve existing forms and their submissions.
`GetForms(...), GetFormByID(...), GetAnswer(...)`
- **Form Deadline**: Set a precise time range during which a form can be interacted with.

## Field Types
The system supports the following field types:

type|example
-|-
string|`{"label": "Name", "fieldType": "string", "required": true}`
number|`{"label": "Age", "fieldType": "number", "required": true}`
boolean|`{"label": "Is Student?", "fieldType": "boolean", "required": false}`
choice|`{"label": "Favorite Food", "fieldType": "[Pizza|Schnitzel|Burger]", "required": true}`
multi-choice|`{"label": "Hobbies", "fieldType": "{Reading|Swimming|Gaming}", "required": false}`

## Demo

The external repo where the initial development took place and where you can find the frontend is [here](https://github.com/agherasie/gno-forms). 

The web app itself is hosted [here](https://gno-forms.netlify.app/)

And the most recent test4 version of the contract is [forms2](https://test4.gno.land/r/g1w62226g8hykfmtuasvz80rdf0jl6phgxsphh5v/testing/forms2)

---

Screenshots :

<details>
<summary>
<a href="https://test4.gno.land/r/g1w62226g8hykfmtuasvz80rdf0jl6phgxsphh5v/testing/forms2">gnoweb Render()</a>
</summary>
<img width="941" alt="image" src="https://github.com/user-attachments/assets/24b9c17d-b51e-4d0b-ab19-b9bca49c0a89">
</details>

<details>
<summary>
<a href="https://gno-forms.netlify.app/results/0000002/g1w62226g8hykfmtuasvz80rdf0jl6phgxsphh5v">
a form response in the web interface
</a>
</summary>
<img width="539" alt="image" src="https://github.com/user-attachments/assets/b3469545-842f-4030-a4da-1060802c6477">
</details>

<details>
<summary>
<a href="https://gno-forms.netlify.app/create">creating a form in the web interface</a></summary>
<img width="564" alt="image" src="https://github.com/user-attachments/assets/c6e16bd0-0523-47a7-bf09-b3dc7f5d9314">
</details>

---

<details><summary>Contributors' checklist...</summary>

- [X] Added new tests, or not needed, or not feasible
- [X] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [X] Updated the official documentation or not needed
- [X] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [ ] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

Gno2D2 commented:
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)
- [ ] The pull request description provides enough details


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 The pull request was created from a fork (head branch repo: agherasie/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

<details><summary>The pull request description provides enough details</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 Not (🔴 Pull request author is a member of the team: core-contributors)
    └── 🟢 Not (🔴 Pull request author is user: dependabot[bot])

```
### Can be checked by
 - team core-contributors

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

notJoon commented on examples/gno.land/p/demo/forms/create.gno:
```diff
@@ -0,0 +1,105 @@
+	fields := make([]Field, fieldsCount)
+
+	// Parsing the json submission to create the gno data structures
+	for i := 0; i < fieldsCount; i++ {
+		field, err := node.GetIndex(i)
+		if err != nil {
+			return "", errInvalidJson
+		}
+
+		labelNode, err := field.GetKey("label")
+		if err != nil {
+			return "", errInvalidJson
+		}
+		fieldTypeNode, err := field.GetKey("fieldType")
+		if err != nil {
+			return "", errInvalidJson
+		}
+		requiredNode, err := field.GetKey("required")
+		if err != nil {
+			return "", errInvalidJson
+		}
+
+		label, err := labelNode.GetString()
+		if err != nil {
+			return "", errInvalidJson
+		}
+		fieldType, err := fieldTypeNode.GetString()
+		if err != nil {
+			return "", errInvalidJson
+		}
+		required, err := requiredNode.GetBool()
+		if err != nil {
+			return "", errInvalidJson
+		}
+
+		fields[i] = CreateField(label, fieldType, required)
+	}
```
In this case, it can be handled thought method chaing using `MustKey` and `MustT` (where `T` is an aribitary type). Although it causes a panic instead of error, this won't occur if the JSON is in the correct form.

```go
for i := 0; i < fieldsCount; i++ {
    field := node.MustIndex(i)
    
    fields[i] = CreateField(
        field.MustKey("label").MustString(),
        field.MustKey("fieldType").MustString(),
        field.MustKey("required").MustBool(),
    )
}
```

notJoon commented on examples/gno.land/p/demo/forms/validate.gno:
```diff
@@ -0,0 +1,150 @@
+}
+
+// ValidateAnswers checks if the given answers are valid for the given fields
+func ValidateAnswers(answers string, fields []Field) bool {
```
It would be good to extract the field validation and choice type checking part into separate functions.

notJoon commented on examples/gno.land/p/demo/forms/validate.gno:
```diff
@@ -0,0 +1,150 @@
+			if validateNumberField(answer, field) == false {
+				return false
+			}
+		} else if field.FieldType[0] == '{' && field.FieldType[len(field.FieldType)-1] == '}' {
+			if validateMultiChoiceField(answer, field) == false {
+				return false
+			}
+		} else if field.FieldType[0] == '[' && field.FieldType[len(field.FieldType)-1] == ']' {
+			if validateChoiceField(answer, field) == false {
+				return false
+			}
```
How do you think about using `strings.HasPrefix` and `strings.HasSuffix` functions? These are much safer than direct index checking.

Plus, you can use `!validateMultiChoiceField(answer, field)` instead of `validateMultiChoiceField(answer, field) == false` to make it more idiomatic.

notJoon commented on examples/gno.land/r/demo/forms/forms.gno:
```diff
@@ -0,0 +1,145 @@
+	return string(encoded)
+}
+
+func GetFormByID(id string) string {
```
When creating JSON package, I made a wrong choice to create object in this complicated way. Now using the builder pattern would allow for a relatively concise approach. 

Additionally, you can make it even more concise by creating pre-defined JSON generation functions for each type and calling them. For example, the basic form building part is being repeated in multiple places, but if you apply the builder pattern metioned above and create it as a separate function, you can modify like this:

```go
func writeBaseFormInfo(builder *json.NodeBuilder, form *Form) *json.NodeBuilder {
    return builder.WriteString("id", form.ID)
        .WriteString("owner", form.Owner.String())
        .WriteString("title", form.Title)
        .WriteString("description", form.Description)
        .WriteString("createdAt", form.CreatedAt.Format("2006-01-02 15:04:05"))
}
```
Then you can call this function at the appropriate location.
```go
func GetFormByID(id string) string {
    form, err := db.GetForm(id)
    if err != nil {
        panic(err)
    }

    builder := json.Builder()
    
    writeBaseFormInfo(builder, form) // <- Called here

    // ...
    encoded, err := json.Marshal(builder.Node()) // <- use `Node()` to finising build
    if err != nil {
        panic(err)
    }

    return string(encoded)
}
```

notJoon reviewed (changes requested):
I've left a few comments. Other than that, looks fine.

agherasie commented:
> I've left a few comments. Other than that, looks fine.

Hi @notJoon thanks so much for your suggestions ! 
I've addressed your comments in various commits, let me know what you think about the current state of the code !

thehowl commented:
@notJoon let's keep the "triage pending" while you folks are still reviewing it ;)

notJoon reviewed (approved):
LGTM 👍 

removing the `review/triage-pending` flag because the review is complete.

leohhhn commented on examples/gno.land/p/demo/forms/README.md:
I believe this should live in a `doc.gno` file :)

leohhhn commented on examples/gno.land/p/demo/forms/create.gno:
```diff
@@ -0,0 +1,80 @@
+// CreateForm creates a new form with the given parameters
+func (db *FormDB) CreateForm(title string, description string, openAt string, closeAt string, data string) (string, error) {
+	// Parsing the dates
+	const dateFormat = "2006-01-02T15:04:05Z"
```
This can actually be defined at top level.

leohhhn commented on examples/gno.land/p/demo/forms/errors.gno:
```diff
@@ -0,0 +1,15 @@
+
+import "errors"
+
+var (
+	errNoOpenDate       = errors.New("Form has no open date")
+	errNoCloseDate      = errors.New("Form has no close date")
+	errInvalidJson      = errors.New("Invalid JSON")
+	errInvalidDate      = errors.New("Invalid date")
+	errFormNotFound     = errors.New("Form not found")
+	errAnswerNotFound   = errors.New("Answer not found")
+	errAlreadySubmitted = errors.New("You already submitted this form")
+	errFormClosed       = errors.New("Form is closed")
+	errInvalidAnswers   = errors.New("Invalid answers")
+)
```
Should these maybe be exported? Do you expect your users to need to use these errors?

leohhhn reviewed (approved):
Just check the last few comments, otherwise looks good. 

You should consider moving this to your own namespace, and also calling `r/leon/hof.Register()` inside the realm, in `init()` so that it shows up on the Hall of Fame :)

agherasie commented:
> @agherasie did you have a chance to check the latest comments?

Hi, sorry for the delay ! I tried to quickly address the comments in my latest commits, however I am running into a problem with the tests. Not sure what it's due to yet ! 
I'm currently traveling so I'm short on time but I'll make sure to finish up everything that's left before the end of the month hopefully, as well as try to get the corresponding r/forms webapp working again since a lot has changed with adena since I made it last summer ! 
That way, once the PR is merged it will be fully functional alongside the companion dapp.

thehowl reviewed (approved):
code quality could be improved, but I'd avoid this PR going stale again so merging seeing other approvals.

thehowl commented:
The CI is failing because of #3240, I think we either need to remove the hof registration or fix that issue...

leohhhn commented:
Ah, I see the PR also needs updating to the latest `std` changes. @agherasie could you work through this and we can merge?

agherasie commented:
Apologies again for the delay, I've been very busy this month as I've been traveling a lot !
I'll make this PR my top priority this weekend :)

> code quality could be improved

@thehowl feel free to add comments where you think I could improve the code if you have the time; as of tomorrow I will be a lot quicker to respond and make changes. The earlier requested changes on this PR have been very insightful for me since I'm pretty inexperienced with go/gno and I was able to learn a lot, so I am open to any further suggestions on how to improve the realm !

Thanks everyone for your help, I'm really excited to finally merging it soon !

thehowl commented:
Let's merge this after #3854, @agherasie please add back the hof registration, it should work after we merge that :)
